File: 102.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ret = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        
        queue.offer(root);
    
        while(!queue.isEmpty()) {
            int len = queue.size();
            List<Integer> subList = new ArrayList<>();
            for(int i = 0; i < len; i++) {
                TreeNode node = queue.poll();   
                if(node != null) {
                    queue.offer(node.left);
                    queue.offer(node.right);
                    subList.add(node.val);
                }
            }
            if(subList.size() > 0)
                ret.add(subList);
        }
        return ret;
    }
}

File: cpp/leetcode/102.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (root == NULL) return res;
        res.push_back({root->val});
        vector<TreeNode *> list = {root};
        while(list.size() != 0) {
            vector<int> vec;
            list = helper(list);
            if (list.size() == 0) break;
            for (auto it: list) {
                vec.push_back(it->val);
            }
            res.push_back(vec);
        }
        return res;
    }
    
    vector<TreeNode *> helper(vector<TreeNode *> level) {
        vector<TreeNode *> list;
        for(auto it : level) {
            if (it->left != NULL) list.push_back(it->left);
            if (it->right != NULL) list.push_back(it->right);
        }
        return list;
    }
};
